File: delete_and_earn_dp.py
# Time Complexity : O(N) + O(M) when N is number of elements in the matrix and M is maximum number in the matrix
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        maxRange = max(nums)
        
        earnings = [0 for _ in range(maxRange + 1)]
        for n in nums:
            earnings[n] += n

        choose = earnings[0]
        notChoose = earnings[1]
        
        for i in range(2, len(earnings)):
            temp = choose
            choose = max(choose, notChoose)
            notChoose = temp + earnings[i]
            
        return max(choose, notChoose)

File: delete-and-earn-dp.cpp
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes

#include <iostream>
#include <vector>
#include <unordered_map>
  
using namespace std; 

class Solution {
public:
    int deleteAndEarn(std::vector<int>& nums) {
        std::unordered_map<int, int> count;
        int maxNum = 0;
        for (int num : nums) {
            count[num]++;
            maxNum = std::max(maxNum, num);
        }

        std::vector<int> dp(maxNum + 1);
        dp[1] = count[1];

        for (int i = 2; i <= maxNum; ++i) {
            dp[i] = std::max(dp[i - 1], dp[i - 2] + count[i] * i);
        }

        return dp[maxNum];
    }
}